### ISSUE TITLE
Is this the correct way to pass Haiku model for evaluation?

### ISSUE THREAD

**Issue by bigyankarki** [deepmind/dm-haiku#510]
Minimal working code:
```
# function to train model on dataset
def train(dataset, batch_size, epochs, lr, momentum, weight_decay):
    # convert from strucutred array to jax array
    state = dataset[0]
    y_policy = dataset[1]
    y_value = dataset[2]
    state = jnp.array([state[i] for i in range(len(state))], dtype=jnp.float32)
    y_policy = jnp.array([y_policy[i] for i in range(len(y_policy))], dtype=jnp.float32)
    y_value = jnp.array([y_value[i] for i in range(len(y_value))], dtype=jnp.float32)

    # initialize model
    rng = jax.random.PRNGKey(0)
    model, params, st = init_model(rng)

     # define optimizer
    optimizer = optax.adamw(learning_rate=lr, b1=momentum, b2=0.999, eps=1e-8, weight_decay=weight_decay)

    # initialize optimizer
    opt_state = optimizer.init(params)
    

    # define loss function
    def loss_fn(params, batch):
        input = batch[0].reshape((batch_size, 4, 4, 1))
        y_predict_value, _st = model.apply(params=params, state=st, x=input, rng=rng)

        y_predict_value = y_predict_value.flatten() # flatten y_predict_label
        y_value = batch[2].flatten() # flatten y_label

        # policy_loss = jnp.mean(optax.softmax_cross_entropy(labels=batch[1], logits=y_predict_policy))
        value_loss = jnp.mean(optax.l2_loss(y_value, y_predict_value))
        return value_loss

    # define update function
    @jax.jit
    def update(params, opt_state, batch):
        loss, grads = jax.value_and_grad(loss_fn)(params, batch)
        updates, opt_state = optimizer.update(grads, opt_state, params)
        params = optax.apply_updates(params, updates)
        return params, opt_state, loss

    print_list = ['epoch', 'Step', 'Loss']
    logging.info(50 * "=")
    logging.info('{:^7} {:^7} {:^8}'.format(*print_list))
    logging.info(50 * "=")
    # print the above list
    print(50 * "=")
    print('{:^7} {:^7} {:^8}'.format(*print_list))
    print(50 * "=")

    # train model
    for epoch in range(10):
        start_time = time.time()
        for i in range(0, dataset.shape[1], batch_size):
            batch_state = state[i:i+batch_size]
            batch_y_policy = y_policy[i:i+batch_size]
            batch_y_value = y_value[i:i+batch_size]

            # if batch size is not equal to batch_size, skip the batch. Happens at the end of the dataset
            if batch_state.shape[0] != batch_size:
                break
            
            params, opt_state, loss = update(params, opt_state, (batch_state, batch_y_policy, batch_y_value))
            
            if (i/batch_size) % 100 == 0:
                log(epoch, i/batch_size, loss)

        end_time = time.time()

        evaluate(model, params, st, rng)
        print('Time: {}'.format(end_time - start_time))
        logging.info("Time: {}".format(end_time - start_time))

    # return trained model
    return model, params, st
```
For evaluation
```
# function to evaluate model on dataset
def evaluate(model, params, st, rng):
    # load test dataset
    test_dataset = np.load('test_set.npy', allow_pickle=True)
    print('Loaded test dataset with {} samples'.format(test_dataset.shape[1]))

    # batch the dataset
    batch_size = test_dataset.shape[1] # a single batch

    # convert from strucutred array to jax array
    state = test_dataset[0]
    y_policy = test_dataset[1]
    y_value = test_dataset[2]
    state = jnp.array([state[i] for i in range(len(state))], dtype=jnp.float32)
    y_policy = jnp.array([y_policy[i] for i in range(len(y_policy))], dtype=jnp.float32)
    y_value = jnp.array([y_value[i] for i in range(len(y_value))], dtype=jnp.float32)

    # evaluate model
    predict_value, _st = model.apply(params=params, state=st, x=state[:batch_size].reshape(batch_size, 4, 4, 1), rng=rng)
    value_loss = jnp.mean(optax.l2_loss(y_value, predict_value))

    # calculate accuracy
    y_predict_value = predict_value.flatten() # flatten y_predict_label
    y_predict_value = jnp.where(y_predict_value <= 0, -1, 1)  # set y_predict_label to -1 if value is <=0  or else 1
    y_value = y_value.flatten() # flatten y_label

    # calculate accuracy
    accuracy = jnp.mean(jnp.where(y_value == y_predict_value, 1, 0))
    print('Accuracy: {}'.format(accuracy))
    logging.info("accuracy in test set: {}".format(accuracy))
    print('Value loss: {}'.format(value_loss))
    logging.info("value loss: {}".format(value_loss))
```

Initializing the Haiku model
```
# function to do forward pass on model
def _forward_fn(x):
    model = AlphaZeroModel(num_filters=256, training=True)
    return model(x)

# function to initialize Haiku model with rng key and input shape
def init_model(rng):
    init_fn = hk.transform_with_state(_forward_fn)
    input_shape = jax.random.normal(rng, (1, 4, 4, 1))
    params, st = init_fn.init(rng, input_shape)
    return init_fn, params, st
```

Input is a 16*1 array representing a state of a board.  Output(y_predict_value) is an outcome, 1 for win and -1 for loss. But during training it's an output from a tanh function of AlphaZero modal. I am just trying to do supervised learning, mapping states to correct outcome. However, although loss seems to be decreasing over time, accuracy is stable at each epoch. I suspect updated parameters is not being passed on to the model during evaluation. Am I passing my parameters correctly in the above code?

**Comment by bigyankarki** [comment]
@LenaMartens I am tagging you from [here](https://github.com/google/jax/discussions/12056#discussioncomment-3459552)

**Comment by bigyankarki** [comment]
```
2022-08-23 15:41:01,352  epoch   Step     Loss  
2022-08-23 15:41:01,352 ==================================================
2022-08-23 15:41:02,716       0     0.0   0.5680
2022-08-23 15:41:33,659       0   100.0   0.6797
2022-08-23 15:42:05,587       0   200.0   0.7500
2022-08-23 15:42:39,580       0   300.0   0.6953
2022-08-23 15:43:12,400       0   400.0   0.7031
2022-08-23 15:43:44,410       0   500.0   0.6328
2022-08-23 15:44:17,089       0   600.0   0.8203
2022-08-23 15:44:48,998       0   700.0   0.7578
2022-08-23 15:45:21,534       0   800.0   0.7266
2022-08-23 15:45:54,228       0   900.0   0.7266
2022-08-23 15:46:46,561 accuracy in test set: 0.6483993530273438
2022-08-23 15:46:46,563 value loss: 0.7032013535499573
2022-08-23 15:46:46,751 Time: 303.14147901535034
2022-08-23 15:46:47,110       1     0.0   0.6719
2022-08-23 15:47:19,287       1   100.0   0.6797
2022-08-23 15:47:52,851       1   200.0   0.7500
2022-08-23 15:48:25,756       1   300.0   0.6953
2022-08-23 15:48:57,571       1   400.0   0.7031
2022-08-23 15:49:30,870       1   500.0   0.6328
2022-08-23 15:50:03,639       1   600.0   0.8203
2022-08-23 15:50:35,683       1   700.0   0.7578
2022-08-23 15:51:07,768       1   800.0   0.7266
2022-08-23 15:51:40,489       1   900.0   0.7266
2022-08-23 15:52:33,334 accuracy in test set: 0.6483993530273438
2022-08-23 15:52:33,335 value loss: 0.7032013535499573
2022-08-23 15:52:33,485 Time: 303.97859477996826
2022-08-23 15:52:33,877       2     0.0   0.6719
2022-08-23 15:53:06,222       2   100.0   0.6797
2022-08-23 15:53:39,334       2   200.0   0.7500
2022-08-23 15:54:11,063       2   300.0   0.6953
2022-08-23 15:54:42,918       2   400.0   0.7031
2022-08-23 15:55:15,249       2   500.0   0.6328
2022-08-23 15:55:47,406       2   600.0   0.8203
2022-08-23 15:56:20,319       2   700.0   0.7578
2022-08-23 15:56:52,048       2   800.0   0.7266
```
It seems that loss repeats after the first step. This means that updated params is not being used?

**Comment by LenaMartens** [comment]
Are the values of your `params` changing? Could it be that the loss function/learning rate is not providing a significant update to your `params`? I would maybe inspect the gradients computed by the update function and make sure they're not 0.

This would explain why the loss never changes. The training loss changes, but I think only because you're evaluating it on different samples of the training set, which is why the loss values reset every epoch. This might just be an issue with the loss function/hyperparameters and a non-training model, could you check the gradients or params before/after the update?

**Comment by bigyankarki** [comment]
Hmm, I think you are right. It's probably the loss function not providing a significant update to params. I just checked updates, and params (before and after the update), and the updates provided after the first few steps are essentially in e-08 order. Here is a log of every 100 steps: https://pastebin.com/fAngsdwt.

I could tune a few hyperparameters: learning rate, momentum, and weight decay. I have tried a few learning rates, but the result is almost the same. 

I am doing this as a sanity check that the AlphaZero model is working. So, I made a dataset and just tried to map the input to outputs. Although the AlphaZero model could be overkill for this simple case, it should be able to learn in my opinion.
